sopel/coretasks.py

```diff
@@ -1250,7 +1264,43 @@ def auth_proceed(bot, trigger):
             bot.write(('AUTHENTICATE', '*'))
             return
 
-    # TODO: Implement SCRAM challenges
+    elif mech == "SCRAM-SHA-256":
+        if trigger.args[0] == "+":
+            bot._scram_client = ScramClient([mech], sasl_username, sasl_password)
+            client_first = bot._scram_client.get_client_first()
+            LOGGER.info("Sending SASL SCRAM client first")
+            send_authenticate(bot, client_first)
+        elif bot._scram_client.stage == ScramClientStage.get_client_first:
+            try:
+                server_first = base64.b64decode(trigger.args[0]).decode("utf-8")
+                bot._scram_client.set_server_first(server_first)
+            except (BinasciiError, KeyError, ScramException) as e:
+                LOGGER.error("SASL SCRAM server_first failed: %r", e)
+                bot.write(("AUTHENTICATE", "*"))
+                return
+            if bot._scram_client.iterations < 4096:
+                LOGGER.warning(
+                    "SASL SCRAM iteration count is insecure, continuing anyway"
+                )
+            elif bot._scram_client.iterations >= 4_000_000:
+                LOGGER.warning(
+                    "SASL SCRAM iteration count is very high, this will be slow..."
+                )
+            client_final = bot._scram_client.get_client_final()
+            LOGGER.info("Sending SASL SCRAM client final")
+            send_authenticate(bot, client_final)
+        elif bot._scram_client.stage == ScramClientStage.get_client_final:
+            try:
+                server_final = base64.b64decode(trigger.args[0]).decode("utf-8")
+                bot._scram_client.set_server_final(server_final)
+            except (BinasciiError, KeyError, ScramException) as e:
+                LOGGER.error("SASL SCRAM server_final failed: %r", e)
+                bot.write(("AUTHENTICATE", "*"))
+                return
+            LOGGER.info("SASL SCRAM succeeded")
+            bot.write(("AUTHENTICATE", "+"))
+            bot._scram_client = None
```

Comment:
Looking at this, I find myself wondering how much of these 30 lines could be pushed into `ScramClient`, so that the `elif` clauses could be folded away behind something roughly like `else: bot._scram_client.proceed()`

I'm not sure if it's really worth pushing across the class boundary, but maybe it would still make sense to define a separate method on the bot.